### PR TITLE
Exposing Kubeflow over load balancer

### LINKED ISSUE #67: How to setup LB endpoint for vanilla Kubeflow deployment
**Is your feature request related to a problem? Please describe.**
@AlexandreBrown on #platform-aws slack

_we have vanilla Kubeflow manifest 1.4 installed on EKS.
We used kubectl port-forward for development and now we would like to expose kubeflow so that we can perform inference requests outside the cluster (without port-forward).
We are wondering which doc to use,_

**Describe the solution you'd like**

Public endpoint to access KF dashboard with dex login

### PR DESCRIPTION
**Pull request by surajkota**
**Which issue is resolved by this Pull Request:**
Resolves #67 
- Automated script, doc and e2e tests for the feature
- Moved relevant content from Cognito README to load balancer guide
- Reduced the no. of steps for setting up load balancer controller

**Testing:**
- Unit tests: Will create a separate PR. Waiting for #156
- [x] e2e tests pass [test-run.log](https://github.com/awslabs/kubeflow-manifests/files/8449203/test-run.txt)
- added test for lb integration in e2e tests
- Manually tested readme instructions
 
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awsconfigs/common/istio-ingress/overlays/https/params.env [hunk 1]
```diff
@@ -0,0 +1 @@
+certArn=
```
Comment (goswamig) [inline awsconfigs/common/istio-ingress/overlays/https/params.env:1]: Can we add a comment here on purpose of this variable? 
I understand the README.md will have the details but adding a line here doesn't hurt.

Comment (goswamig) [reply awsconfigs/common/istio-ingress/overlays/https/params.env:1]: I would also suggest to create another overlay for http.

File: docs/deployment/add-ons/load-balancer/README.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# Exposing Kubeflow over Load Balancer
+
+This tutorial shows how to expose Kubeflow over a load balancer on AWS.
+
+## Background
+
+Kubeflow does not offer a generic solution for connecting to Kubeflow over a load balancer because this process is highly dependent on your environment/cloud provider. On AWS, we use the [AWS Load Balancer controller](https://kubernetes-sigs.github.io/aws-load-balancer-controller/) which satisfies the Kubernetes [Ingress resource](https://kubernetes.io/docs/concepts/services-networking/ingress/) to create an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB). When you create a Kubernetes `Ingress`, an ALB is provisioned that load balances application traffic.
+
+In order to connect to Kubeflow using a LoadBalancer, we need to setup HTTPS. The reason is that many of the Kubeflow web apps (e.g., Tensorboard Web App, Jupyter Web App, Katib UI) use [Secure Cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#restrict_access_to_cookies), so accessing Kubeflow with HTTP over a non-localhost domain does not work.
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:9]: What about other components like pipeline, serving ? Do they also not support HTTP ?

File: docs/deployment/add-ons/load-balancer/README.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+To secure the traffic and use HTTPS, we must associate a Secure Sockets Layer/Transport Layer Security (SSL/TLS) certificate with the load balancer. [AWS Certificate Manager](https://aws.amazon.com/certificate-manager/) is a service that lets you easily provision, manage, and deploy public and private Secure Sockets Layer/Transport Layer Security (SSL/TLS) certificates for use with AWS services and your internal connected resources. To create a certificate for use with the load balancer, you must specify a domain name i.e. certificates cannot be created for ALB DNS. You can register your domain using any domain service provider such as [Route53](https://aws.amazon.com/route53/), GoDoddy etc.
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:11]: Can you link the reference doc that mentioned this limitation ?

`certificates cannot be created for ALB DNS`

File: docs/deployment/add-ons/load-balancer/README.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+- configure the parameters for [load balancer controller](../../../../awsconfigs/common/aws-alb-ingress-controller/base/params.env) with the cluster name 
+- create an IAM role with [these permissions](../../../../awsconfigs/infra_configs/iam_alb_ingress_policy.json) for the load balancer controller to use via a service account(`alb-ingress-controller` service account in `kubeflow` namespace). You can use the [eksctl create iamserviceaccount](https://eksctl.io/usage/iamserviceaccounts/) command for this.
+
+You can then use `kubectl` and `kustomize` to build and install the components specified in this [kustomize](./kustomization.yaml) file.
+
+After installing the components, the script will wait for the ingress managed load balancer to be created and update the subdomain (`platform.example.com`) with a CNAME entry for `kubeflow.platform.example.com` pointing to the load balancer DNS. You can see the load balancer DNS by running `kubectl get ingress -n istio-system istio-ingress`.
+
+To summarize, what the script does or what you need to if you are doing it manually:
+    - Create a custom domain to host Kubeflow
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:42]: `domain and subdomain`

File: docs/deployment/add-ons/load-balancer/README.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    - Create TLS certificates for the domain using ACM
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:43]: `domain and subdomain`

File: docs/deployment/add-ons/load-balancer/README.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    - Configure the ingress and load balancer controller
+    - Build manifests and install the required components
+    - Update the domain with ALB address
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:46]: `subdomain`

File: docs/deployment/add-ons/load-balancer/README.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+1. Installed the tools mentioned in [prerequisite section of this](../../vanilla/README.md#prerequisites) document on the client machine.
+
+## Creating the load balancer
+
+To make it easy to create the load balancer, you can use the [script](#run-the-script) provided in the instructions below. Following is a brief overview of what happens when you run the script:
+### Domain and Certificate creation
+
+As explained in the background section, you need a registered domain and TLS certificate to use HTTPS with load balancer. Since your top level domain(e.g. `example.com`) could have been registered at any service provider, for uniformity and taking advantage of the integration provided between Route53, ACM and Application Load Balancer, you will create a separate [sudomain](https://en.wikipedia.org/wiki/Subdomain) (e.g. `platform.example.com`) to host Kubeflow and a corresponding hosted zone in Route53 to route traffic for this subdomain. To get TLS support, you will need certificates for both the root domain(`*.example.com`) and subdomain(`*.platform.example.com`).
+
+### Ingress and Load Balancer Controller configuration
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:29]: If I am assuming it right, you need to write precise command that user has to run by following this guide.

File: docs/deployment/add-ons/load-balancer/README.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    - Create TLS certificates for the domain using ACM
+    - Configure the ingress and load balancer controller
+    - Build manifests and install the required components
+    - Update the domain with ALB address
+
+### Run the script
+
+1. Install dependencies for the script
+    ```
+    cd tests/e2e
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:52]: why we are asking to moving to tests directory if its user facing README?

File: docs/deployment/add-ons/load-balancer/README.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    pip install -r requirements.txt
+    ```
+1. Substitute values in `tests/e2e/utils/load_balancer/config.yaml`.
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:55]: I am lost here, are we writing this steps for developer to test this feature ? or we writing this for end user ?

File: docs/deployment/add-ons/load-balancer/README.md (hunk 8 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md:135]: This is a limitation we are facing with using python right now to run these as part of tests, validate script dependencies and avoid any code duplication. We took this trade off so the scripts can be run in same way in tests or if user runs it

We can investigate how to enhance this but it is out of scope for this PR.

File: docs/deployment/add-ons/load-balancer/README.md (hunk 9 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md:138]: answered above

File: docs/deployment/cognito/README.md [hunk 10]
```diff
@@ -92,16 +89,20 @@ In this section, we will be creating certificate to enable TLS authentication at
...
-    1. ![cognito-active-domain](./images/cognito-active-domain.png)
-    2. When it’s created, it will return the `Alias target` CloudFront address.
-        1. Screenshot of the CloudFront URL for Cognito Domain name:
-            1. ![cognito-domain-cloudfront-url](./images/cognito-domain-cloudfront-url.png)
-        2.  Create a new record of type `A` for `auth.platform.example.com` with the value of the CloudFront URL.
-            1. Select the `alias` toggle and select Alias to Cloudfront distribution for creating the record
-            2. ![cognito-domain-cloudfront-A-record-creating](./images/cognito-domain-cloudfront-A-record-creatin.png)
-            3. Following is a screenshot of the A record for `auth.platform.example.com` in `platform.example.com` hosted zone:
-                1. ![cognito-domain-cloudfront-A-record-created](./images/cognito-domain-cloudfront-A-record-created.png)
+1. Add a custom domain to the user pool. In order to add a custom domain to your user pool, you need specify a domain name, and provide a certificate managed with AWS Certificate Manager (ACM).
```
[the remark this answers is not in the document]
Comment (surajkota) [reply docs/deployment/cognito/README.md:74]: I was trying to make the domain and cert creation generic to reuse it across different guides and hence moved the cognito specific things. I think the new structure in this revision will make it clear

File: docs/deployment/cognito/README.md [hunk 11]
```diff
@@ -13,7 +13,7 @@ This guide assumes that you have:
...
     - [kustomize version 3.2.0](https://github.com/kubernetes-sigs/kustomize/releases/tag/v3.2.0) - A command line tool to customize Kubernetes objects through a kustomization file.
-      - :warning: Kubeflow 1.3.0 is not compatible with the latest versions of of kustomize 4.x. This is due to changes in the order resources are sorted and printed. Please see [kubernetes-sigs/kustomize#3794](https://github.com/kubernetes-sigs/kustomize/issues/3794) and [kubeflow/manifests#1797](https://github.com/kubeflow/manifests/issues/1797). We know this is not ideal and are working with the upstream kustomize team to add support for the latest versions of kustomize as soon as we can.
+      - :warning: Kubeflow is not compatible with the latest versions of of kustomize 4.x. This is due to changes in the order resources are sorted and printed. Please see [kubernetes-sigs/kustomize#3794](https://github.com/kubernetes-sigs/kustomize/issues/3794) and [kubeflow/manifests#1797](https://github.com/kubeflow/manifests/issues/1797). We know this is not ideal and are working with the upstream kustomize team to add support for the latest versions of kustomize as soon as we can.
```
[the remark this answers is not in the document]
Comment (surajkota) [reply docs/deployment/cognito/README.md:16]: Will do in future

File: docs/deployment/add-ons/load-balancer/README.md (hunk 7 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md]: Addressed this comment in the latest version of guide. Previously, I was trying to only keep automated script the primary way to do this.

File: tests/e2e/utils/load_balancer/lb_resources_cleanup.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        if "kubeflow" in cfg.keys():
+            alb = cfg["kubeflow"].get("alb", None)
+            if alb:
+                subprocess.call(
+                    f"kubectl delete ingress -n istio-system istio-ingress".split()
+                )
+                # load balancer controller does not place a finalizer on the ingress and so deleting the attached load balancer is asynhronous
+                # adding a random wait to allow controller to delete the load balancer
+                # TODO: implement a better check
+                time.sleep(2 * 60)
```
[the remark this answers is not in the document]
Comment (surajkota) [reply tests/e2e/utils/load_balancer/lb_resources_cleanup.py]: Added a better check with exception if deletion does not happen

**Comment by surajkota** [comment]
@goswamig I have made significant changes to the guide from the last review and addressed comments as well. Please take a look

@ryansteakley can you take a look as well?

File: docs/deployment/add-ons/load-balancer/README.md (hunk 8 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md:135]: https://github.com/awslabs/kubeflow-manifests/issues/166

**Comment by goswamig** [comment]
I would suggest to use 

also there are some error, specifically if we are not deleting the resources, over the time they will pilled up.

```
ingress.networking.k8s.io "istio-ingress" deleted
INFO:__main__:d7ee0b0f-istiosystem-istio-2af2-1382882308.ca-central-1.elb.amazonaws.com: waiting for deletion ...
ERROR:e2e.utils.aws.elbv2:Failed to describe load balancer d7ee0b0f-istiosystem-istio-2af2
Traceback (most recent call last):
  File "/home/ubuntu/kubeflow/kubeflow-manifests/tests/e2e/utils/aws/elbv2.py", line 28, in describe
    response = self.elbv2_client.describe_load_balancers(
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 388, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 708, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.LoadBalancerNotFoundException: An error occurred (LoadBalancerNotFound) when calling the DescribeLoadBalancers operation: Load balancers '[d7ee0b0f-istiosystem-istio-2af2]' not found
2022-04-08 06:27:46 [ℹ]  eksctl version 0.82.0
2022-04-08 06:27:46 [ℹ]  using region ca-central-1
2022-04-08 06:27:46 [ℹ]  1 iamserviceaccount (kubeflow/alb-ingress-controller) was included (based on the include/exclude rules)
2022-04-08 06:27:46 [ℹ]  1 task: { 
    2 sequential sub-tasks: { 
        delete IAM role for serviceaccount "kubeflow/alb-ingress-controller" [async],
        delete serviceaccount "kubeflow/alb-ingress-controller",
    } }2022-04-08 06:27:46 [ℹ]  will delete stack "eksctl-surakota-dev-10-addon-iamserviceaccount-kubeflow-alb-ingress-controller"
2022-04-08 06:27:46 [ℹ]  deleted serviceaccount "kubeflow/alb-ingress-controller"
ERROR:e2e.utils.aws.iam:failed to delete iam policy arn:aws:iam::169544399729:policy/alb_ingress_controller_surakota-dev-10gyche6w5po
Traceback (most recent call last):
  File "/home/ubuntu/kubeflow/kubeflow-manifests/tests/e2e/utils/aws/iam.py", line 51, in delete
    self.iam_client.delete_policy(
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 388, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ubuntu/anaconda3/envs/kf/lib/python3.8/site-packages/botocore/client.py", line 708, in _make_api_call
    raise error_class(parsed_response, operation_name)

```

File: docs/deployment/add-ons/load-balancer/README.md (hunk 2 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md:13]: looks like kserve webapp too based on: https://www.kubeflow.org/docs/external-add-ons/kserve/webapp/#port-forwarding

**Comment by surajkota** [comment]
@goswamig The two errors you see:
1. failed to describe alb - script waits after deleting ingress for the ALB to be deleted. The error log appears when ALB is successfully deleted
2. for IAM policy error - there is also because of a retry. There is a delay in IAM service for the attached role deletion to reflect in the policy. It fails initially after the role is deleted but as soon as policy is up to date, the deletion is successful
```
INFO:e2e.utils.aws.iam:deleted iam policy arn:aws:iam::169544399729:policy/alb_ingress_controller_surakota-dev-10gyche6w5po
```

File: docs/deployment/add-ons/load-balancer/README.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        ```
+        kubectl config current-context
+        aws eks describe-cluster --name $CLUSTER_NAME
+        ```
+1. Verify the current directory is the root of the repository by running the `pwd` command. The output should be `<path/to/kubeflow-manifests>` directory
+
+
+## Create Load Balancer
+
+To make it easy to create the load balancer, you can use the [script provided in this section](#automated-script). If you prefer to use rhe automated scripts, you need to only execute the steps in the [automated script section](#automated-script). Read the following sections in this guide to understand what happens when you run the script or execute all the steps if you prefer to do it manually/hands-on.
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:37]: typo:  If you prefer to use `rhe` automated scripts

File: docs/deployment/add-ons/load-balancer/README.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+            - ```
+                export TAG_VALUE=<>
+                export CLUSTER_SUBNET_IDS=$(aws ec2 describe-subnets --region $CLUSTER_REGION --filters Name=tag:alpha.eksctl.io/cluster-name,Values=$CLUSTER_NAME --output json | jq -r '.Subnets[].SubnetId')
+                for i in "${CLUSTER_SUBNET_IDS[@]}"
+                do
+                    aws ec2 create-tags --resources ${i} --tags Key=kubernetes.io/cluster/${CLUSTER_NAME},Value=${TAG_VALUE}
+                done
+                ```
+        1. `kubernetes.io/role/internal-elb`. Add this tag only to private subnet
+        1. `kubernetes.io/role/elb`. Add this tag only to public subnet.
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:95]: I think there are multiple private and public subnet, right ?
if so,
subnet -> subnets

File: docs/deployment/add-ons/load-balancer/README.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+kustomize build docs/deployment/add-ons/load-balancer | kubectl apply -f -
+```
+
+### Update the domain with ALB address
+
+1. Check if ALB is provisioned. It takes around 3-5 minutes
+    1. ```
+        kubectl get ingress -n istio-system istio-ingress
+        NAME            CLASS    HOSTS   ADDRESS                                                                  PORTS   AGE
+        istio-ingress   <none>   *       ebde55ee-istiosystem-istio-2af2-1100502020.us-west-2.elb.amazonaws.com   80      15d
```
Comment (goswamig) [inline docs/deployment/add-ons/load-balancer/README.md:123]: would suggest put dummy address.

File: tests/e2e/utils/load_balancer/setup_load_balancer.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+
+    alb = ElasticLoadBalancingV2(dns=alb_dns, region=region)
+    def callback():
+        assert alb.describe()["State"]["Code"] == expected_status
+
+    wait_for(callback)
+
+
+def create_ingress():
+    kubectl_apply_kustomize(path=common.LB_KUSTOMIZE_PATH)
```
Comment (goswamig) [inline tests/e2e/utils/load_balancer/setup_load_balancer.py:198]: where do i see this file ? the only file in path `LB_KUSTOMIZE_PATH = "../../docs/deployment/add-ons/load-balancer/"` is README.md am I missing something here?

File: tests/e2e/utils/load_balancer/setup_load_balancer.py [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+
+
+def dns_update(
+    region: str, cluster_name: str, subdomain_hosted_zone: Route53HostedZone
+) -> str:
+
+    wait_for_alb_dns(cluster_name, region)
+    ingress = get_ingress(cluster_name, region)
+    alb_dns = ingress["status"]["loadBalancer"]["ingress"][0]["hostname"]
+    wait_for_alb_status(alb_dns, region)
```
Comment (goswamig) [inline tests/e2e/utils/load_balancer/setup_load_balancer.py:208]: makes sense, the dns gets populated first before marking the status active.

**Review by goswamig — COMMENTED** [review]
almost there,  few minor comments and questions.

File: docs/deployment/add-ons/load-balancer/README.md (hunk 15 above)
Comment (surajkota) [reply docs/deployment/add-ons/load-balancer/README.md:123]: this is dummy address, replace few chars with xxx to make it more explicit

File: tests/e2e/utils/load_balancer/setup_load_balancer.py (hunk 16 above)
Comment (surajkota) [reply tests/e2e/utils/load_balancer/setup_load_balancer.py:198]: https://github.com/surajkota/kubeflow-manifests/blob/lb_dex/docs/deployment/add-ons/load-balancer/kustomization.yaml


its part of this PR, can you check again?

**Review by ryansteakley — APPROVED** [review]

**Review by goswamig — APPROVED** [review]

**Pull request merged**